processing/src/main/java/org/apache/druid/data/input/impl/RetryingInputStream.java

```diff
@@ -69,23 +69,62 @@ public RetryingInputStream(
       Predicate<Throwable> retryCondition,
       @Nullable Integer maxTries
   ) throws IOException
+  {
+    this(object, objectOpenFunction, retryCondition, maxTries, true);
+  }
+
+  @VisibleForTesting
+  RetryingInputStream(
+      T object,
+      ObjectOpenFunction<T> objectOpenFunction,
+      Predicate<Throwable> retryCondition,
+      @Nullable Integer maxTries,
+      boolean doWait
+  ) throws IOException
   {
     this.object = Preconditions.checkNotNull(object, "object");
     this.objectOpenFunction = Preconditions.checkNotNull(objectOpenFunction, "objectOpenFunction");
     this.retryCondition = Preconditions.checkNotNull(retryCondition, "retryCondition");
     this.maxTries = maxTries == null ? RetryUtils.DEFAULT_MAX_TRIES : maxTries;
-    this.delegate = new CountingInputStream(objectOpenFunction.open(object));
-    this.doWait = true;
+    this.doWait = doWait;
+    this.startOffset = 0;
 
     if (this.maxTries <= 1) {
       throw new IAE("maxTries must be greater than 1");
     }
+    openWithRetry();
   }
 
   private void openIfNeeded() throws IOException
   {
     if (delegate == null) {
-      delegate = new CountingInputStream(objectOpenFunction.open(object, startOffset));
+      openWithRetry();
+    }
+  }
+
+  private void openWithRetry() throws IOException
```

Comment:
nit: lets pass startOffset to openWithRetry as a function parameter since this method never mutates the offset. 
Then you will not need line 90: this.startOffset = 0 and the code becomes cleaner.